### PR TITLE
add handshakeTimeout, bound chunkLength, comments

### PR DESCRIPTION
**Pull request by ebuchman**
- ReadFull hangs forever unless we set a timeout on the conn (not sure this matters for sconn.Read(), but it does for shareEphPubKey

- sconn.Read was missing a bounds check such that a malicious `chunkLength` would crash the node

- there's also a question in there about the size of the buffer used for `shareAuthSignature`

Unfortunately, I couldn't think of an easy way to implement tests for my fixes (I tested both manually).

### CODE REVIEW

**Comment by ebuchman** [comment]
crypto byte arrays (pubkey, priv key, sig for ed25519) are all fixed length now. 
I added a very basic test (ie. a "it doesnt panic" test) for fixed length arrays, will still add more.
NOTE: moving to fixed length breaks the previous pubkey->addr map because we pubkey.Address() binary encodes the pubkey before hashing

File: binary/reflect.go [hunk 1]
```diff
@@ -864,8 +865,9 @@ func writeReflectJSON(rv reflect.Value, rt reflect.Type, w io.Writer, n *int64,
 		length := rt.Len()
 		if elemRt.Kind() == reflect.Uint8 {
 			// Special case: Bytearray
-			bytearray := rv.Interface()
-			WriteTo([]byte(Fmt("\"%X\"", bytearray)), w, n, err)
```
Comment (jaekwon) [inline binary/reflect.go]: was this necessary?

Comment (ebuchman) [inline binary/reflect.go]: I didn't like it. But with the old code it seemed to be hex encoding the call to eg pubKey.String()

Comment (ebuchman) [inline binary/reflect.go]: do you have a better way to do this?

**Pull request merged**